### DISCUSSION TITLE
Generate graphics plots as part of WE2E test

### DISCUSSION THREAD

**Discussion by BruceKropp-Raytheon** [ufs-community/ufs-srweather-app#741] [Ideas]
The tutorial describes how to manually generate plots, and in order to automate this, we would need a simple way to enable it to happen with a workflow E2E test run.
https://srw-ug.readthedocs.io/en/text-tutorial/Tutorial.html#tutorials

**Comment by EdwardSnyder-NOAA** [comment]
Good morning Bruce!

We have one [WE2E test](https://github.com/ufs-community/ufs-srweather-app/blob/develop/tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional_plot.yaml) that tests the plotting task. Are there specific WE2E tests you would like to have the plotting task enabled for? 

In @mkavulich issue #587 on overhauling and consolidating WE2E tests, he mentioned that plotting might not be suitable for WE2E testing. Maybe he can chime in on how to automate the WE2E plotting task.

In the meantime, you could use the sed command to insert the plotting task into the WE2E test config files found under the ufs-srweather-app/tests/WE2E/test_configs/grids_extrn_mdls_suites_(community|nco) directories.
`sed -i '<line number>i rocoto:' /path/to/config yaml`
`sed -i '<line number>i \ \ tasks:' /path/to/config yaml`
`sed -i '<line number>i \ \ \ \ taskgroups: '\''{{ ["parm/wflow/prep.yaml", "parm/wflow/coldstart.yaml", "parm/wflow/post.yaml", "parm/wflow/plot.yaml"]|include }}'\''' /path/to/config yaml`

Thanks!

Eddie Snyder | Advanced User Support (AUS) Team

**Reply by mkavulich** [reply]
@EdwardSnyder-NOAA RE: my comments in #587, I was only throwing out the idea of adding plotting tests without knowing the state of that capability. I didn't mean to imply that the plotting task was somehow "unsuitable" for running in the workflow.

@BruceKropp-Raytheon As Eddie pointed out, we already have a workflow end-to-end test that demonstrates the plotting capability. Sed seems like overkill for modifying these files, simply inserting the following lines to any `config.yaml` file will enable plotting tasks:

```
rocoto:
  tasks:
    taskgroups: '{{ ["parm/wflow/prep.yaml", "parm/wflow/coldstart.yaml", "parm/wflow/post.yaml", "parm/wflow/plot.yaml"]|include }}'
```

If the "taskgroups" definition already exists, adding the `"parm/wflow/plot.yaml"` entry to the list will do it. Sorry I can't point you to documentation for the taskgroups definitions, this is a new revision to how the workflow is generated (#676) and documentation is still being developed.

Note that plotting is currently hard-coded for a fixed CONUS domain, though the logic does exist in the plotting script (scripts/exregional_plot_allvars.py) that would allow for a custom domain if you are comfortable modifying python code, see lines 417 and 518.

**Reply by BruceKropp-Raytheon** [reply]
Thanks Eddie and Mike
I have tried the one E2E test that is configured to generate a plot, but did not see it produce a plot. So maybe that feature isn't working as expected.
I have also tried the sed changes suggested here, with little success.
I am willing to try again, if these are indeed the right paths to take, but it does seem cumbersome when trying to automate this.
The idea is to add plotting as a future validation gate for CI/CD PR builds, so it would be nice to be able to choose any of the experiment results to test that plotting still works with the PR change.
Ideally we need a reliable method that can be scripted into say a jenkinsfile at some point in the future.

**Reply by mkavulich** [reply]
@BruceKropp-Raytheon I just checked the latest run of the plotting test I have and it looks like there is a bug in the latest workflow that is preventing the correct plotting task from running. I will open an issue about this.

With the issues you mentioned in mind, it may be a good idea to start manually adding plotting tasks to more WE2E tests. They are very lightweight tasks so I don't think there's any resource-based reason to refrain from adding plotting to more tasks. I don't think this is the kind of thing that should be handled in a CI/CD pipeline, but rather be a feature of the tests themselves (especially since not all tests will produce model output). The CI/CD pipeline *could* serve as a way to check it a plotting task was run for a given task, and save those image files somewhere. That would be a nice feature to have.

**Reply by BruceKropp-Raytheon** [reply]
Totally agree, Mike!
Thanks for checking this. And great idea for having it part of WE2E tests, since we already have that as a validation stage.
Bruce

**Reply by mkavulich** [reply]
Okay, I have opened issue #742, it will be fixed in PR #732 when that is merged (I confirmed the easy fix worked).

**Reply by EdwardSnyder-NOAA** [reply]
@mkavulich - my apologies for misinterpreting your comments in #587, which is why I suggested doing sed as I thought we weren't going to modify any WE2E tests. Given that's not the case, should we go ahead and add plotting capabilities to a number of commonly used WE2E tests?

**Reply by mkavulich** [reply]
I think that would be a good idea. It would be really nice if we could support plotting for all/arbitrary domains, but I'm not super familiar with the state of the plotting capability right now, so I'm not sure how much work it would be to implement that. So maybe just the CONUS domains for now?

**Reply by EdwardSnyder-NOAA** [reply]
The plot_allvars task does plotting for CONUS and regional domains. The regional domain is defined by the latitude and longitude found in the post-processed prslev and natlev grib2 files. 

The workflow defaults to the CONUS domain. To enable it for the regional domain, add the below code to the end of the config WE2E yaml. Make sure the `rocoto` step we mentioned earlier is in the config WE2E yaml as well.
````
task_plot_allvars:
  PLOT_DOMAINS: ["regional"]
````
I looked through our WE2E tests and think that we should add plotting capabilities to the config yamls listed below. Each one of these files are in one or more coverage, comprehensive, or fundamental list, meaning that each testing suite will have at least one experiment that runs the plot_allvars task.
````
grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
grid_RRFS_NA_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_2017_gfdlmp
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_WoFS_v0
````
Let me know what you think, and I can open a PR for these changes.

**Reply by BruceKropp-Raytheon** [reply]
Great! Let's add this.

Would the default configs be ready-to-go with rocoto for the usual WE2E coverage tests that CM runs?

**Reply by EdwardSnyder-NOAA** [reply]
I created a [draft PR](https://github.com/ufs-community/ufs-srweather-app/pull/777) that adds plotting capabilities to a few WE2E tests. Now every comprehensive, coverage, and fundamental suite will have at least one experiment that tests the graphics plotting capabilities. I did end up using a whole different list of tests. To answer your question, yes the selected configs have been renamed and the rocoto section has been added to them.

**Comment by gspetro-NOAA** [comment]
PR https://github.com/ufs-community/ufs-srweather-app/pull/777 solved this problem.
